### PR TITLE
Use a program counter to control shaders' flow

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
This should more precise and allow ssy/sync instructions to be implemented.
It may cause performance regressions on shader-limited GPUs.

Thanks Subv for the suggestion.

### CODE REVIEW

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs [hunk 1]
```diff
@@ -435,6 +437,37 @@ private void PrintMain()
             SB.AppendLine("}");
         }
 
+        private void PrintProgram(ShaderIrBlock[] Blocks, string Name)
+        {
+            string Ident1 = IdentationStr;
```
Comment (gdkchan) [inline Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: `const string`?

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs [hunk 2]
```diff
@@ -435,6 +437,37 @@ private void PrintMain()
...
+            string Ident2 = Ident1 + IdentationStr;
+            string Ident3 = Ident2 + IdentationStr;
+            string Ident4 = Ident3 + IdentationStr;
+
+            SB.AppendLine(Ident1 + "pc = " + GetBlockPosition(Blocks[0]) + ";");
+
+            SB.AppendLine(Ident1 + "do {");
```
Comment (gdkchan) [inline Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: I think the empty lines between `SB.AppendLine` are not necessary.

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs [hunk 3]
```diff
@@ -435,6 +437,37 @@ private void PrintMain()
...
+            SB.AppendLine(Ident2 + "switch (pc) {");
+
+            foreach (ShaderIrBlock Block in Blocks)
+            {
+                string FunctionName = Block.Position.ToString("x8");
+
+                SB.AppendLine(Ident3 + "case 0x" + FunctionName + ": pc = " + Name + "_" + FunctionName + "(); break;");
+            }
+
+            SB.AppendLine(Ident3 + "default:");
```
Comment (gdkchan) [inline Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: I think the empty lines between `SB.AppendLine` are not necessary.

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs [hunk 4]
```diff
@@ -961,7 +913,7 @@ private string GetClampuExpr(ShaderIrOp Op)
 
         private string GetCnumExpr(ShaderIrOp Op) => GetUnaryCall(Op, "!isnan");
 
-        private string GetExitExpr(ShaderIrOp Op) => "return";
+        private string GetExitExpr(ShaderIrOp Op) => "return 0";
```
Comment (gdkchan) [inline Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: Maybe use `return 0u` for consistency with the other one?

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs [hunk 5]
```diff
@@ -468,193 +501,112 @@ private void PrintAttrToOutput(string Identation = IdentationStr)
...
+                            TailBranch = true;
+                            break;
                     }
                 }
-            }
 
-            return Tail;
+                if (!TailBranch)
+                {
+                    SB.AppendLine(IdentationStr + "// Jump to next block");
```
Comment (gdkchan) [inline Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: Imo this comment is not necessary.

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs [hunk 6]
```diff
@@ -468,193 +501,112 @@ private void PrintAttrToOutput(string Identation = IdentationStr)
...
-            return Blocks.FirstOrDefault(x => x.Position == Block.EndPosition);
-        }
+            if (Nodes.Length > 0)
+            {
+                ShaderIrNode Last = Nodes[Nodes.Length - 1];
 
-        private ShaderIrBlock GetLoopTailBlock(ShaderIrBlock LoopHead)
-        {
-            ShaderIrBlock Tail = null;
+                bool TailBranch = false;
```
Comment (gdkchan) [inline Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: I think that a better name is `UnconditionalBranch`. Althrough Exit and Kil are not branches, so maybe `UnconditionalFlowChange`?

File: Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs (hunk 4 above)
Comment (ReinUsesLisp) [reply Ryujinx.Graphics/Gal/Shader/GlslDecompiler.cs]: I didn't add them for style. glslangValidator for glsl 4.1 doesn't allow implicit conversions from `int` to `uint`.

**Comment by ReinUsesLisp** [comment]
Closing in favor of #382.

**Pull request closed without merging**